File: ambeth/Ambeth.Service/ambeth/service/AbstractOfflineServiceUrlProvider.cs
using De.Osthus.Ambeth.Config;
using De.Osthus.Ambeth.Ioc;
using De.Osthus.Ambeth.Ioc.Extendable;
using De.Osthus.Ambeth.Util;
using System;
using System.ComponentModel;
using System.Threading;

namespace De.Osthus.Ambeth.Service
{
    public abstract class AbstractOfflineServiceUrlProvider : IServiceUrlProvider, IOfflineListenerExtendable, IInitializingBean, INotifyPropertyChanged
    {
        public virtual event PropertyChangedEventHandler PropertyChanged;

        protected bool isOffline;

        [Property(ServiceConfigurationConstants.OfflineMode, DefaultValue = "false")]
        public virtual bool IsOffline
        {
            get
            {
                return isOffline;
            }
            set
            {
                if (isOffline == value)
                {
                    return;
                }
                isOffline = value;
                ThreadPool.QueueUserWorkItem(delegate(Object handle)
                {
                    IsOfflineChanged();
                    SyncContext.Post(delegate(Object handle2)
                    {
                        if (PropertyChanged != null)
                        {
                            PropertyChanged.Invoke(this, new PropertyChangedEventArgs("IsOffline"));
                        }
                    }, null);
                }, null);
            }
        }

        protected readonly IExtendableContainer<IOfflineListener> offlineListeners = new DefaultExtendableContainer<IOfflineListener>("offlineListener");

        public SynchronizationContext SyncContext { get; set; }

        public virtual void AfterPropertiesSet()
        {
            ParamChecker.AssertParamNotNull(SyncContext, "SyncContext");
        }

        public void LockForRestart(bool offlineAfterRestart)
        {
            IOfflineListener[] listeners = offlineListeners.GetExtensions();

            foreach (IOfflineListener offlineListener in listeners)
            {
                if (offlineAfterRestart)
                {
                    offlineListener.BeginOffline();
                }
                else
                {
                    offlineListener.BeginOnline();
                }
            }
            foreach (IOfflineListener offlineListener in listeners)
            {
                if (offlineAfterRestart)
                {
                    offlineListener.HandleOffline();
                }
                else
                {
                    offlineListener.HandleOnline();
                }
            }
        }

        protected virtual void IsOfflineChanged()
        {
            IOfflineListener[] listeners = offlineListeners.GetExtensions();

            foreach (IOfflineListener offlineListener in listeners)
            {
                if (isOffline)
                {
                    offlineListener.BeginOffline();
                }
                else
                {
                    offlineListener.BeginOnline();
                }
            }
            foreach (IOfflineListener offlineListener in listeners)
            {
                if (isOffline)
                {
                    offlineListener.HandleOffline();
                }
                else
                {
                    offlineListener.HandleOnline();
                }
            }
            foreach (IOfflineListener offlineListener in listeners)
            {
                if (isOffline)
                {
                    offlineListener.EndOffline();
                }
                else
                {
                    offlineListener.EndOnline();
                }
            }
        }

        public abstract String GetServiceURL(Type serviceInterface, String serviceName);

        public virtual void AddOfflineListener(IOfflineListener offlineListener)
        {
            offlineListeners.Register(offlineListener);
        }

        public virtual void RemoveOfflineListener(IOfflineListener offlineListener)
        {
            offlineListeners.Unregister(offlineListener);
        }
    }
}

File: jambeth/jambeth-service/src/main/java/de/osthus/ambeth/service/AbstractOfflineServiceUrlProvider.java
package de.osthus.ambeth.service;

import de.osthus.ambeth.config.Property;
import de.osthus.ambeth.config.ServiceConfigurationConstants;
import de.osthus.ambeth.ioc.DefaultExtendableContainer;
import de.osthus.ambeth.ioc.IInitializingBean;
import de.osthus.ambeth.ioc.extendable.IExtendableContainer;

public abstract class AbstractOfflineServiceUrlProvider implements IServiceUrlProvider, IOfflineListenerExtendable, IInitializingBean
{
	protected boolean isOffline;

	@Override
	public boolean isOffline()
	{
		return isOffline;
	}

	@Override
	@Property(name = ServiceConfigurationConstants.OfflineMode, defaultValue = "false")
	public void setOffline(boolean isOffline)
	{
		if (this.isOffline == isOffline)
		{
			return;
		}
		this.isOffline = isOffline;

		isOfflineChanged();
	}

	protected final IExtendableContainer<IOfflineListener> offlineListeners = new DefaultExtendableContainer<IOfflineListener>(IOfflineListener.class,
			"offlineListener");

	@Override
	public void afterPropertiesSet()
	{
		// Intended blank
	}

	@Override
	public void lockForRestart(boolean offlineAfterRestart)
	{
		IOfflineListener[] listeners = offlineListeners.getExtensions();

		for (IOfflineListener offlineListener : listeners)
		{
			if (offlineAfterRestart)
			{
				offlineListener.beginOffline();
			}
			else
			{
				offlineListener.beginOnline();
			}
		}
		for (IOfflineListener offlineListener : listeners)
		{
			if (offlineAfterRestart)
			{
				offlineListener.handleOffline();
			}
			else
			{
				offlineListener.handleOnline();
			}
		}
	}

	protected void isOfflineChanged()
	{
		IOfflineListener[] listeners = offlineListeners.getExtensions();

		for (IOfflineListener offlineListener : listeners)
		{
			if (isOffline)
			{
				offlineListener.beginOffline();
			}
			else
			{
				offlineListener.beginOnline();
			}
		}
		for (IOfflineListener offlineListener : listeners)
		{
			if (isOffline)
			{
				offlineListener.handleOffline();
			}
			else
			{
				offlineListener.handleOnline();
			}
		}
		for (IOfflineListener offlineListener : listeners)
		{
			if (isOffline)
			{
				offlineListener.endOffline();
			}
			else
			{
				offlineListener.endOnline();
			}
		}
	}

	@Override
	public abstract String getServiceURL(Class<?> serviceInterface, String serviceName);

	@Override
	public void addOfflineListener(IOfflineListener offlineListener)
	{
		offlineListeners.register(offlineListener);
	}

	@Override
	public void removeOfflineListener(IOfflineListener offlineListener)
	{
		offlineListeners.unregister(offlineListener);
	}
}